homeassistant/components/number/__init__.py

```diff
@@ -397,6 +402,17 @@ def unit_of_measurement(self) -> str | None:
             and self.device_class == NumberDeviceClass.TEMPERATURE
         ):
             return self.hass.config.units.temperature_unit
+        if (
+            native_unit_of_measurement
+            in (UnitOfTemperatureDelta.CELSIUS, UnitOfTemperatureDelta.FAHRENHEIT)
+            and self.device_class == NumberDeviceClass.TEMPERATURE_DELTA
+        ):
+            # UnitSystem.temperature_unit is either °F or °C so we need to map it
+            # to the corresponding delta unit
+            return {
+                UnitOfTemperature.CELSIUS: UnitOfTemperatureDelta.CELSIUS,
+                UnitOfTemperature.FAHRENHEIT: UnitOfTemperatureDelta.FAHRENHEIT,
+            }.get(self.hass.config.units.temperature_unit, native_unit_of_measurement)
```

Comment:
The special case for absolute temperatures here is for historical reasons, why do we need to add this conversion also for temperature deltas?